utils/bitfield/src/lib.rs

```diff
@@ -0,0 +1,453 @@
+    Decoded(BitVec),
+}
+
+impl Default for BitField {
+    fn default() -> Self {
+        Self::Decoded(BitVec::new())
+    }
+}
+
+impl BitField {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Generates a new bitfield with a slice of all indexes to set.
+    pub fn new_from_set(set_bits: &[u64]) -> Self {
+        let mut vec = match set_bits.iter().max() {
+            Some(&max) => {
+                let mut vec = BitVec::with_capacity(max as usize + 1);
+                vec.resize(max as usize + 1, false);
+                vec
+            }
+            None => return Self::Decoded(BitVec::new()),
+        };
+
+        // Set all bits in bitfield
+        for b in set_bits {
+            vec.set(*b as usize, true);
+        }
+
+        Self::Decoded(vec)
+    }
+
+    /// Sets bit at bit index provided
+    pub fn set(&mut self, bit: u64) {
+        match self {
+            BitField::Encoded { set, unset, .. } => {
+                unset.remove(&bit);
+                set.insert(bit);
+            }
+            BitField::Decoded(bv) => {
+                let index = bit as usize;
+                if bv.len() <= index {
+                    bv.resize(index + 1, false);
+                }
+                bv.set(index, true);
+            }
+        }
+    }
+
+    /// Removes bit at bit index provided
+    pub fn unset(&mut self, bit: u64) {
+        match self {
+            BitField::Encoded { set, unset, .. } => {
+                set.remove(&bit);
+                unset.insert(bit);
+            }
+            BitField::Decoded(bv) => {
+                let index = bit as usize;
+                if bv.len() < index {
+                    return;
+                }
+                bv.set(index, false);
+            }
+        }
+    }
+
+    /// Gets the bit at the given index.
+    // TODO this probably should not require mut self and RLE decode bits
+    pub fn get(&mut self, index: u64) -> Result<bool> {
+        match self {
+            BitField::Encoded { set, unset, .. } => {
+                if set.contains(&index) {
+                    return Ok(true);
+                }
+
+                if unset.contains(&index) {
+                    return Ok(false);
+                }
+
+                // Check in encoded for the given bit
+                // This can be changed to not flush changes
+                if let Some(true) = self.as_mut_flushed()?.get(index as usize) {
+                    Ok(true)
+                } else {
+                    Ok(false)
+                }
+            }
+            BitField::Decoded(bv) => {
+                if let Some(true) = bv.get(index as usize) {
+                    Ok(true)
+                } else {
+                    Ok(false)
+                }
+            }
+        }
+    }
+
+    /// Retrieves the index of the first set bit, and error if invalid encoding or no bits set.
```

Comment:
I mean, only unwrapping for the Rust builtin operations which have equivalent functions to return the result if needed (those are setup for convenience as they can't error as long as the bitfield is decoded as invalid RLE encoding). 

To be clear, in this case it requires the type to be decoded before checking, not encoded. I was planning on opening an issue after this PR comes in detailing the optimization of this to rely on iterating over the encoded bits, but I can do that now after this

Reply:
The trade-off here is that doing a single pass over the RLE+-encoded to check its validity could have a negative impact on performance, if I understand correctly? I get that you're being cautious but the cost of this check should absolutely be dwarfed by the actual encoding/decoding operations. So to be clear, I'm talking about a dedicated function that only checks whether the encoded bitvec is valid RLE+ or not, and not about completely decoding every incoming RLE+ bitvec as it comes in.

I don't mind having a look at this later, actually sounds really fun, but I don't see any downsides at all

Reply:
No it's not just to check validity, it's to be able to get and set bits in O(1) time after it's decoded. if you just check validity, you still need to iterate over the bits to get the first decoded set bit in this case, or whether a bit is set or not in the indexing case, which is why it should come with realistic benchmarks

Reply:
That's not really what I was suggesting at all, I'm not talking about decoding it vs not — just that whenever `BitField::Decoded` is constructed, it would ideally be checked for validity first because then we know that the bitvec is valid and from that point flushing can't fail anymore. No need to do that in this PR, but it's not really something we needs encoded vs decoded benchmarks for

Reply:
Every Decoded bitfield is valid though, only encoded bitfields can be invalid (and since enum as of now can't really separate the function into one with a result, but the decoded is impossible to fail as there is nothing to flush)

Reply:
Sorry, I meant the construction of `BitField::Encoded`, not `Decoded`. Preventing `BitField::Encoded` from being constructed with an invalid value would make flushing infallible

Reply:
ah yes but then we would error in different places than the go implementation, and there would be a consensus fault if that edge case is hit (happy to explain this in more detail)